tests/unit/lib/upload_manager_test.py

```diff
@@ -23,110 +29,113 @@ def update_bundle(self, *args, **kwargs):
+
+    def test_fileobj_single_tar_gz_with_dsstore_should_not_simplify_archive_2(self):
+        """If the user included three files, README, README2, and .DS_Store, in the archive,
+        the archive should not be simplified because we have more than one file in the archive.
+        """
+        source = os.path.join(self.temp_dir, 'source_dir')
+        os.mkdir(source)
+        self.write_string_to_file('testing', os.path.join(source, 'README'))
+        self.write_string_to_file('testing', os.path.join(source, 'README2'))
+        self.write_string_to_file('testing', os.path.join(source, '.DS_Store'))
+        self.do_upload([('source.tar.gz', tar_gzip_directory(source))])
+        self.assertEqual(
+            ['.DS_Store', 'README', 'README2'], sorted(os.listdir(self.bundle_location))
+        )
+
+    def mock_url_sources(self, fileobj, ext=""):
+        """Returns a URL that is mocked to return the contents of fileobj.
+        The URL will end in the extension "ext", if given.
+        """
+        url = f"https://codalab/contents{ext}"
+        size = len(fileobj.read())
+        fileobj.seek(0)
+        urllib.request.urlopen = MagicMock()
+        urllib.request.urlopen.return_value = addinfourl(fileobj, {"content-length": size}, url)
+        return [url]
+
+    def test_url_single(self):
+        self.do_upload(self.mock_url_sources(BytesIO(b'hello world')))
+        self.check_file_contains_string(self.bundle_location, 'hello world')
+
+    def test_url_tar_gz(self):
+        source = os.path.join(self.temp_dir, 'source_dir')
+        os.mkdir(source)
+        self.write_string_to_file('testing', os.path.join(source, 'file1'))
+        self.write_string_to_file('testing', os.path.join(source, 'file2'))
+        self.do_upload(
+            self.mock_url_sources(BytesIO(tar_gzip_directory(source).read()), ext=".tar.gz")
+        )
+        self.assertIn('file2', os.listdir(self.bundle_location))
+
     def test_multiple_sources(self):
-        self.do_upload([('source1', BytesIO(b'testing1')), ('source2', BytesIO(b'testing2'))])
-        self.assertEqual(['source1', 'source2'], sorted(os.listdir(self.bundle_location)))
+        self.do_upload(
+            [
+                ('source1', BytesIO(b'testing1')),
+                ('source2', BytesIO(b'testing2')),
+                'http://alpha.gnu.org/gnu/bc/bc-1.06.95.tar.bz2',
+            ]
+        )
+        self.assertEqual(
+            ['bc-1.06.95', 'source1', 'source2'], sorted(os.listdir(self.bundle_location))
+        )
         self.check_file_contains_string(os.path.join(self.bundle_location, 'source1'), 'testing1')
         self.check_file_contains_string(os.path.join(self.bundle_location, 'source2'), 'testing2')
+        self.assertIn('README', os.listdir(os.path.join(self.bundle_location, 'bc-1.06.95')))
+
+    def test_url_git(self):
+        self.do_upload(['https://github.com/codalab/test'], git=True)
```

Comment:
this url doesn't exist, right? what's the purpose?

Reply:
Actually, it exists -- this test hits the actual network, as I couldn't find an easy way to mock a git repository url. See https://github.com/codalab/test